File: api/src/main/java/bot/inker/ankh/core/api/world/storage/StorageBackend.java
package bot.inker.ankh.core.api.world.storage;

import bot.inker.ankh.core.api.storage.ChunkStorage;

import javax.annotation.Nonnull;
import java.util.List;
import java.util.stream.Stream;

/**
 * StorageBackend store world data, likes block's pos and meta
 */
public interface StorageBackend {
   /**
    * Provide a stream include all ankh-blocks in WorldChunkEmbedded
    */
   Stream<BlockStorageEntry> provide(@Nonnull ChunkStorage worldChunk);

   /**
    * Store ankh-blocks in WorldChunkEmbedded
    */
   void store(@Nonnull ChunkStorage worldChunk, @Nonnull List<BlockStorageEntry> entries);
}

File: src/main/kotlin/bot/inker/ankh/core/world/storage/StorageBackend.kt
package bot.inker.ankh.core.world.storage

import bot.inker.ankh.core.common.entity.WorldChunkEmbedded
import java.util.stream.Stream

/**
 * StorageBackend store world data, likes block's pos and meta
 */
interface StorageBackend {
  /**
   * Provide a stream include all ankh-blocks in WorldChunkEmbedded
   */
  fun provide(worldChunk: WorldChunkEmbedded): Stream<BlockStorageEntry>

  /**
   * Store ankh-blocks in WorldChunkEmbedded
   */
  fun store(worldChunk: WorldChunkEmbedded, entries: List<BlockStorageEntry>)
}